From 1b3a6738e6741949a5aa6f958b004d8c6fba1651 Mon Sep 17 00:00:00 2001
From: Poweruser <poweruser.rs@hotmail.com>
Date: Sat, 21 Mar 2015 06:35:52 +0100
Subject: [PATCH] Dont Ignore Snooper Setting


diff --git a/src/main/java/net/minecraft/server/DedicatedServer.java b/src/main/java/net/minecraft/server/DedicatedServer.java
index d57a796..3616662 100644
--- a/src/main/java/net/minecraft/server/DedicatedServer.java
+++ b/src/main/java/net/minecraft/server/DedicatedServer.java
@@ -36,6 +36,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
     private boolean generateStructures;
     private WorldSettings.EnumGamemode r;
     private boolean s;
+    protected boolean snooperEnabled; // MinetickMod
 
     // CraftBukkit start - Signature changed
     public DedicatedServer(joptsimple.OptionSet options) {
@@ -226,7 +227,7 @@ public class DedicatedServer extends MinecraftServer implements IMinecraftServer
                 this.aA();
                 this.getEnableCommandBlock();
                 this.p();
-                this.getSnooperEnabled();
+                this.snooperEnabled = this.getSnooperEnabled(); // MinetickMod
                 this.aJ();
                 this.c(this.propertyManager.getInt("max-build-height", 256));
                 this.c((this.getMaxBuildHeight() + 8) / 16 * 16);
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index b18aef9..82af764 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -649,11 +649,11 @@ public abstract class MinecraftServer implements Runnable, ICommandListener, IAs
         this.h[this.ticks % 100] = System.nanoTime() - i;
         this.methodProfiler.b();
         this.methodProfiler.a("snooper");
-        if (!this.n.d() && this.ticks > 100) {
+        if (this.snooperEnabled && !this.n.d() && this.ticks > 100) { // MinetickMod
             this.n.a();
         }
 
-        if (this.ticks % 6000 == 0) {
+        if (this.snooperEnabled && this.ticks % 6000 == 0) { // MinetickMod
             this.n.b();
         }
 
-- 
1.8.3.msysgit.0

