From 4aa1fdeb3782449803313b4b8493f1d8c5e1cb60 Mon Sep 17 00:00:00 2001
From: Poweruser <poweruser.rs@hotmail.com>
Date: Sat, 21 Mar 2015 10:32:12 +0100
Subject: [PATCH] Fix Disappearing Leash Hitches


diff --git a/src/main/java/net/minecraft/server/EntityInsentient.java b/src/main/java/net/minecraft/server/EntityInsentient.java
index c8156aa..769979f 100644
--- a/src/main/java/net/minecraft/server/EntityInsentient.java
+++ b/src/main/java/net/minecraft/server/EntityInsentient.java
@@ -268,6 +268,11 @@ public abstract class EntityInsentient extends EntityLiving {
 
             nbttagcompound.set("Leash", nbttagcompound1);
         }
+        // MinetickMod start
+        else if(this.bq != null) {
+            nbttagcompound.set("Leash", this.bq);
+        }
+        // MinetickMod end
 
         if (this.ce()) {
             nbttagcompound.setBoolean("NoAI", this.ce());
@@ -868,11 +873,12 @@ public abstract class EntityInsentient extends EntityLiving {
                     EntityLiving entityliving = (EntityLiving) iterator.next();
 
                     if (entityliving.getUniqueID().equals(uuid)) {
-                        this.bp = entityliving;
+                        this.setLeashHolder(entityliving, true); // MinetickMod
                         break;
                     }
                 }
-            } else if (this.bq.hasKeyOfType("X", 99) && this.bq.hasKeyOfType("Y", 99) && this.bq.hasKeyOfType("Z", 99)) {
+            }
+            if (this.bp == null && this.bq.hasKeyOfType("X", 99) && this.bq.hasKeyOfType("Y", 99) && this.bq.hasKeyOfType("Z", 99)) { // MinetickMod
                 BlockPosition blockposition = new BlockPosition(this.bq.getInt("X"), this.bq.getInt("Y"), this.bq.getInt("Z"));
                 EntityLeash entityleash = EntityLeash.b(this.world, blockposition);
 
@@ -880,8 +886,9 @@ public abstract class EntityInsentient extends EntityLiving {
                     entityleash = EntityLeash.a(this.world, blockposition);
                 }
 
-                this.bp = entityleash;
-            } else {
+                this.setLeashHolder(entityleash, true); // MinetickMod
+            }
+            if (this.bp == null) { // MinetickMod
                 this.world.getServer().getPluginManager().callEvent(new EntityUnleashEvent(this.getBukkitEntity(), UnleashReason.UNKNOWN)); // CraftBukkit
                 this.unleash(false, true);
             }
-- 
1.8.3.msysgit.0

