From 654541e516df8e9903a7a31ec13790ccf8c45ce3 Mon Sep 17 00:00:00 2001
From: Poweruser <poweruser.rs@hotmail.com>
Date: Sat, 21 Mar 2015 14:39:47 +0100
Subject: [PATCH] Fix Insane Nether Portal Lag


diff --git a/src/main/java/net/minecraft/server/PortalTravelAgent.java b/src/main/java/net/minecraft/server/PortalTravelAgent.java
index 4d4b683..acb607b 100644
--- a/src/main/java/net/minecraft/server/PortalTravelAgent.java
+++ b/src/main/java/net/minecraft/server/PortalTravelAgent.java
@@ -94,7 +94,13 @@ public class PortalTravelAgent {
 
     public boolean b(Entity entity, float f) {
         // CraftBukkit start - Modularize portal search process and entity teleportation
-        BlockPosition found = this.findPortal(entity.locX, entity.locY, entity.locZ, 128);
+        // MinetickMod start
+        BlockPosition found = this.findPortal(entity.locX, entity.locY, entity.locZ, 10);
+        if (found == null) {
+            found = this.findPortal(entity.locX, entity.locY, entity.locZ, 128);
+        }
+        // MinetickMod end
+
         if (found == null) {
             return false;
         }
@@ -132,28 +138,34 @@ public class PortalTravelAgent {
             flag1 = false;
         } else {
             BlockPosition blockposition = new BlockPosition(x, y, z);
-
-            for (int l = -128; l <= 128; ++l) {
-                BlockPosition blockposition1;
-
-                for (int i1 = -128; i1 <= 128; ++i1) {
-                    for (BlockPosition blockposition2 = blockposition.a(l, this.a.V() - 1 - blockposition.getY(), i1); blockposition2.getY() >= 0; blockposition2 = blockposition1) {
-                        blockposition1 = blockposition2.down();
-                        if (this.a.getType(blockposition2).getBlock() == Blocks.PORTAL) {
-                            while (this.a.getType(blockposition1 = blockposition2.down()).getBlock() == Blocks.PORTAL) {
-                                blockposition2 = blockposition1;
+            // MinetickMod start
+            BlockPosition.MutableBlockPosition mutableblockposition = new BlockPosition.MutableBlockPosition();
+            int range = short1;
+
+            int zOffset = 0, yOffset = 0;
+            for (int xDiff = -range; xDiff <= range; ++xDiff) {
+                zOffset = (zOffset + 1) % 2;
+                for (int zDiff = -range + zOffset; zDiff <= range; zDiff = zDiff + 2) { // skipping every 2nd block in z direction and alternating from row to row in x direction
+                    yOffset = (yOffset + 1) % 3;
+                    for (int yPos = this.a.V() - (1 + yOffset) ; yPos >= 0; yPos = yPos - 3) { // checking only every 3rd block in y direction and alternating in height in each column
+                        mutableblockposition.c(blockposition.getX() + xDiff, yPos, blockposition.getZ() + zDiff);
+                        if (this.a.getType(mutableblockposition).getBlock() == Blocks.PORTAL) {
+                            int lowestCorrectHeight = mutableblockposition.getY();
+                            while (this.a.getType(mutableblockposition.c(mutableblockposition.getX(), mutableblockposition.getY() - 1, mutableblockposition.getZ())).getBlock() == Blocks.PORTAL) {
+                                lowestCorrectHeight = mutableblockposition.getY();
                             }
 
-                            double d1 = blockposition2.i(blockposition);
+                            double d1 = mutableblockposition.c(mutableblockposition.getX(), lowestCorrectHeight, mutableblockposition.getZ()).i(blockposition);
 
                             if (d0 < 0.0D || d1 < d0) {
                                 d0 = d1;
-                                object = blockposition2;
+                                object = new BlockPosition(mutableblockposition);
                             }
                         }
                     }
                 }
             }
+            // MinetickMod end
         }
 
         if (d0 >= 0.0D) {
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java b/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java
index cae65cf..a7504b9 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftTravelAgent.java
@@ -44,7 +44,12 @@ public class CraftTravelAgent extends PortalTravelAgent implements TravelAgent {
     @Override
     public Location findPortal(Location location) {
         PortalTravelAgent pta = ((CraftWorld) location.getWorld()).getHandle().getTravelAgent();
-        BlockPosition found = pta.findPortal(location.getX(), location.getY(), location.getZ(), this.getSearchRadius());
+        // MinetickMod start
+        BlockPosition found = pta.findPortal(location.getX(), location.getY(), location.getZ(), 10);
+        if (found == null) {
+            found = pta.findPortal(location.getX(), location.getY(), location.getZ(), this.getSearchRadius());
+        }
+        // MinetickMod end
         return found != null ? new Location(location.getWorld(), found.getX(), found.getY(), found.getZ(), location.getYaw(), location.getPitch()) : null;
     }
 
-- 
1.8.3.msysgit.0

