From 6f4ddece2845a48ad18553a3984c4bffcbf605c7 Mon Sep 17 00:00:00 2001
From: Poweruser <poweruser.rs@hotmail.com>
Date: Tue, 7 Jul 2015 08:30:40 +0200
Subject: [PATCH] Allow Disabling of Chunk Generation


diff --git a/src/main/java/de/minetick/MinetickEmptyChunk.java b/src/main/java/de/minetick/MinetickEmptyChunk.java
new file mode 100644
index 0000000..ee06944
--- /dev/null
+++ b/src/main/java/de/minetick/MinetickEmptyChunk.java
@@ -0,0 +1,105 @@
+package de.minetick;
+
+import com.google.common.base.Predicate;
+import java.util.List;
+import java.util.Random;
+
+import net.minecraft.server.AxisAlignedBB;
+import net.minecraft.server.Block;
+import net.minecraft.server.Blocks;
+import net.minecraft.server.BlockPosition;
+import net.minecraft.server.Chunk;
+import net.minecraft.server.Entity;
+import net.minecraft.server.EnumSkyBlock;
+import net.minecraft.server.IChunkProvider;
+import net.minecraft.server.TileEntity;
+import net.minecraft.server.World;
+
+public class MinetickEmptyChunk extends Chunk {
+
+    public MinetickEmptyChunk(World world, int i, int j) {
+        super(world, i, j);
+        this.lit = true;
+        this.done = true;
+    }
+
+    public boolean a(int i, int j) {
+        return i == this.locX && j == this.locZ;
+    }
+
+    public int b(int i, int j) {
+        return 1;
+    }
+
+    public void initLighting() {}
+
+    public Block getType(BlockPosition blockposition) {
+        return Blocks.AIR;
+    }
+
+    public int b(BlockPosition blockposition) {
+        return 255;
+    }
+
+    public int c(BlockPosition blockposition) {
+        return 0;
+    }
+
+    public int getBrightness(EnumSkyBlock enumskyblock, BlockPosition blockposition) {
+        return enumskyblock.c;
+    }
+
+    public void a(EnumSkyBlock enumskyblock, BlockPosition blockposition, int i) {}
+
+    public int a(BlockPosition blockposition, int i) {
+        return 0;
+    }
+
+    public void a(Entity entity) {}
+
+    public void b(Entity entity) {}
+
+    public void a(Entity entity, int i) {}
+
+    public boolean d(BlockPosition blockposition) {
+        return false;
+    }
+
+    public TileEntity a(BlockPosition blockposition, Chunk.EnumTileEntityState chunk_enumtileentitystate) {
+        return null;
+    }
+
+    public void a(TileEntity tileentity) {}
+
+    public void a(BlockPosition blockposition, TileEntity tileentity) {}
+
+    public void e(BlockPosition blockposition) {}
+
+    public void addEntities() {}
+
+    public void removeEntities() {}
+
+    public void e() {}
+
+    public void a(Entity entity, AxisAlignedBB axisalignedbb, List<Entity> list, Predicate<? super Entity> predicate) {}
+
+    public <T extends Entity> void a(Class<? extends T> oclass, AxisAlignedBB axisalignedbb, List<T> list, Predicate<? super T> predicate) {}
+
+    public boolean a(boolean flag) {
+        return false;
+    }
+
+    public Random a(long i) {
+        return new Random(this.getWorld().getSeed() + (long) (this.locX * this.locX * 4987142) + (long) (this.locX * 5947611) + (long) (this.locZ * this.locZ) * 4392871L + (long) (this.locZ * 389711) ^ i);
+    }
+
+    public boolean isEmpty() {
+        return true;
+    }
+
+    public boolean c(int i, int j) {
+        return true;
+    }
+
+    public void loadNearby(IChunkProvider ichunkprovider, IChunkProvider ichunkprovider1, int i, int j) {}
+}
diff --git a/src/main/java/de/minetick/MinetickModConfig.java b/src/main/java/de/minetick/MinetickModConfig.java
index c14a63b..a4d121d 100644
--- a/src/main/java/de/minetick/MinetickModConfig.java
+++ b/src/main/java/de/minetick/MinetickModConfig.java
@@ -41,6 +41,7 @@ public class MinetickModConfig {
     private final Logger logger = LogManager.getLogger();
     private File configFile;
     private FileConfiguration configuration;
+    private HashSet<String> notGeneratingWorlds = new HashSet<String>();
 
     public MinetickModConfig(File configFile) throws IOException, InvalidConfigurationException {
         this.configFile = configFile;
@@ -65,7 +66,7 @@ public class MinetickModConfig {
     }
 
     private void loadConfigContent() {
-
+        this.loadNotGeneratingWorlds(this.getNotGeneratingWorlds());
     }
 
     private FileConfiguration loadConfig(File file) throws IOException, InvalidConfigurationException {
@@ -111,4 +112,19 @@ public class MinetickModConfig {
     public int getProfilerLogInterval() {
         return configuration.getInt("minetickmod.profiler-log-interval", 10);
     }
+
+    public List<String> getNotGeneratingWorlds() {
+        return configuration.getStringList("minetickmod.notGeneratingWorlds");
+    }
+
+    private void loadNotGeneratingWorlds(List<String> worlds) {
+        this.notGeneratingWorlds.clear();
+        for(String w: worlds) {
+            this.notGeneratingWorlds.add(w.toLowerCase());
+        }
+    }
+
+    public boolean doesWorldNotGenerateChunks(String worldName) {
+        return this.notGeneratingWorlds.contains(worldName.toLowerCase());
+    }
 }
diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 1b68736..1edacb6 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -32,8 +32,8 @@ public class Chunk {
     private boolean k;
     public final Map<BlockPosition, TileEntity> tileEntities;
     public final EntitySlice<Entity>[] entitySlices;
-    private boolean done;
-    private boolean lit;
+    protected boolean done; // MinetickMod - protected
+    protected boolean lit; //MinetickMod - protected
     private boolean p;
     private boolean q;
     private boolean r;
diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index a749ecc..c52f29d 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -1,6 +1,7 @@
 package net.minecraft.server;
 
 import com.google.common.collect.Lists;
+
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Collections;
@@ -23,6 +24,11 @@ import org.bukkit.craftbukkit.util.LongObjectHashMap;
 import org.bukkit.event.world.ChunkUnloadEvent;
 // CraftBukkit end
 
+// MinetickMod start
+import de.minetick.MinetickEmptyChunk;
+import de.minetick.MinetickMod;
+// MinetickMod end
+
 public class ChunkProviderServer implements IChunkProvider {
 
     private static final Logger b = LogManager.getLogger();
@@ -34,8 +40,13 @@ public class ChunkProviderServer implements IChunkProvider {
     public LongObjectHashMap<Chunk> chunks = new LongObjectHashMap<Chunk>();
     public WorldServer world;
 
+    // MinetickMod start
+    private MinetickEmptyChunk mtEmptyChunk;
+    // MinetickMod end
+
     public ChunkProviderServer(WorldServer worldserver, IChunkLoader ichunkloader, IChunkProvider ichunkprovider) {
         this.emptyChunk = new EmptyChunk(worldserver, Integer.MIN_VALUE, Integer.MIN_VALUE); // MinetickMod
+        this.mtEmptyChunk = new MinetickEmptyChunk(worldserver, Integer.MIN_VALUE, Integer.MIN_VALUE); // MinetickMod
         this.world = worldserver;
         this.chunkLoader = ichunkloader;
         this.chunkProvider = ichunkprovider;
@@ -136,6 +147,10 @@ public class ChunkProviderServer implements IChunkProvider {
             if (chunk == null) {
                 if (this.chunkProvider == null) {
                     chunk = this.emptyChunk;
+                // MinetickMod start
+                } else if (MinetickMod.getConfig().doesWorldNotGenerateChunks(this.world.getWorld().getName())) {
+                    return this.mtEmptyChunk;
+                // MinetickMod end
                 } else {
                     try {
                         chunk = this.chunkProvider.getOrCreateChunk(i, j);
@@ -194,7 +209,7 @@ public class ChunkProviderServer implements IChunkProvider {
 
         chunk = chunk == null ? (!this.world.ad() && !this.forceChunkLoad ? this.emptyChunk : this.getChunkAt(i, j)) : chunk;
 
-        if (chunk == emptyChunk) return chunk;
+        if (chunk == emptyChunk || chunk == this.mtEmptyChunk) return chunk; // MinetickMod
         if (i != chunk.locX || j != chunk.locZ) {
             b.error("Chunk (" + chunk.locX + ", " + chunk.locZ + ") stored at  (" + i + ", " + j + ") in world '" + world.getWorld().getName() + "'");
             b.error(chunk.getClass().getName());
@@ -230,7 +245,7 @@ public class ChunkProviderServer implements IChunkProvider {
     }
 
     public void saveChunkNOP(Chunk chunk) {
-        if (this.chunkLoader != null) {
+        if (this.chunkLoader != null && !chunk.isEmpty()) { // MinetickMod - dont save empty chunks
             try {
                 this.chunkLoader.b(this.world, chunk);
             } catch (Exception exception) {
@@ -241,7 +256,7 @@ public class ChunkProviderServer implements IChunkProvider {
     }
 
     public void saveChunk(Chunk chunk) {
-        if (this.chunkLoader != null) {
+        if (this.chunkLoader != null && !chunk.isEmpty()) { // MinetickMod - dont save empty chunks
             try {
                 chunk.setLastSaved(this.world.getTime());
                 this.chunkLoader.a(this.world, chunk);
diff --git a/src/main/resources/configurations/minetickmod.yml b/src/main/resources/configurations/minetickmod.yml
index ebf121b..3393dda 100644
--- a/src/main/resources/configurations/minetickmod.yml
+++ b/src/main/resources/configurations/minetickmod.yml
@@ -2,3 +2,4 @@ minetickmod:
     profiler-log-interval: 30
     profiler-write-log: false
     profiler-write-interval: 10
+    notGeneratingWorlds: []
-- 
1.8.3.msysgit.0

