From 71f33e08ff5f834ab8082cf3d44315607c6ff61a Mon Sep 17 00:00:00 2001
From: Poweruser <poweruser.rs@hotmail.com>
Date: Sun, 22 Mar 2015 12:38:51 +0100
Subject: [PATCH] Readd Chunk Access Cache in World


diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 90cdac4..8981a20 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -43,6 +43,18 @@ public class Chunk {
     private int v;
     private ConcurrentLinkedQueue<BlockPosition> w;
 
+    // MinetickMod start
+    private boolean unloaded = false;
+
+    public boolean wasUnloaded() {
+        return this.unloaded;
+    }
+
+    public void markAsUnloaded() {
+        this.unloaded = true;
+    }
+    // MinetickMod end
+
     // CraftBukkit start - Neighbor loaded cache for chunk lighting and entity ticking
     private int neighbors = 0x1 << 12;
 
diff --git a/src/main/java/net/minecraft/server/ChunkProviderServer.java b/src/main/java/net/minecraft/server/ChunkProviderServer.java
index a318574..b399dc0 100644
--- a/src/main/java/net/minecraft/server/ChunkProviderServer.java
+++ b/src/main/java/net/minecraft/server/ChunkProviderServer.java
@@ -364,6 +364,7 @@ public class ChunkProviderServer implements IChunkProvider {
                 if (!event.isCancelled()) {
 
                     if (chunk != null) {
+                        chunk.markAsUnloaded(); // MinetickMod
                         chunk.removeEntities();
                         this.saveChunk(chunk);
                         this.saveChunkNOP(chunk);
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 056799e..0d602ef 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -170,6 +170,10 @@ public abstract class World implements IBlockAccess {
     public boolean instaFall = false;
     private List<Entity> dimensionChangeQueue = Collections.synchronizedList(new ArrayList<Entity>());
     protected boolean cancelHeavyCalculations = false;
+    private Chunk dummyChunk = new EmptyChunk(this, Integer.MIN_VALUE, Integer.MIN_VALUE);
+    private Chunk lastChunkAccessed = dummyChunk;
+    final Object chunkLock = new Object();
+    public ChunkProviderServer chunkProviderServer; // moved here from WorldServer
 
     public void queueEntityForDimensionChange(Entity entity) {
         this.dimensionChangeQueue.add(entity);
@@ -304,8 +308,30 @@ public abstract class World implements IBlockAccess {
         return this.getChunkAt(blockposition.getX() >> 4, blockposition.getZ() >> 4);
     }
 
+    // MinetickMod start
+    private void cacheLastChunkAccess(Chunk foundChunk) {
+        this.lastChunkAccessed = ((foundChunk == null || foundChunk.isEmpty()) ? this.dummyChunk : foundChunk);
+    }
+    // MinetickMod end
+
     public Chunk getChunkAt(int i, int j) {
-        return this.chunkProvider.getOrCreateChunk(i, j);
+        // MinetickMod start
+        Chunk last = this.lastChunkAccessed;
+        if(last.a(i,j) && !last.wasUnloaded()) {
+            return last;
+        } else {
+            Chunk result = null;
+            if(this.chunkProviderServer.isChunkLoaded(i, j)) {
+                result = this.chunkProviderServer.getChunkAt(i, j);
+            } else {
+                synchronized (this.chunkLock) {
+                    result = this.chunkProvider.getOrCreateChunk(i, j);
+                }
+            }
+            this.cacheLastChunkAccess(result);
+            return result;
+        }
+        // MinetickMod end
     }
 
     public boolean setTypeAndData(BlockPosition blockposition, IBlockData iblockdata, int i) {
diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index b1521ad..4e3e879 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -38,7 +38,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
     // private final Set<NextTickListEntry> L = Sets.newHashSet(); // CraftBukkit, PAIL: Rename nextTickListHash
     private final HashTreeSet<NextTickListEntry> M = new HashTreeSet<NextTickListEntry>(); // CraftBukkit - HashTreeSet, PAIL: Rename nextTickList
     private final Map<UUID, Entity> entitiesByUUID = Maps.newHashMap();
-    public ChunkProviderServer chunkProviderServer;
+    //public ChunkProviderServer chunkProviderServer; // MinetickMod - moved to World
     public boolean savingDisabled;
     private boolean O;
     private int emptyTime;
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 9e158e1..117bf16 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -185,6 +185,7 @@ public class CraftWorld implements World {
             save = true;
         }
 
+        chunk.markAsUnloaded(); // MinetickMod
         chunk.removeEntities(); // Always remove entities - even if discarding, need to get them out of world table
 
         if (save && !(chunk instanceof EmptyChunk)) {
-- 
1.8.3.msysgit.0

