From f9f04b4f2338f6df283efea5d9cf956758d10155 Mon Sep 17 00:00:00 2001
From: Poweruser <poweruser.rs@hotmail.com>
Date: Sat, 25 Apr 2015 19:46:36 +0200
Subject: [PATCH] Fix Null Tile Entity Worlds


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 38eb8d3..02b2e9b 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1878,6 +1878,7 @@ public abstract class World implements IBlockAccess {
                     }
                 }
 
+                tileentity.a(this); // MinetickMod
                 this.b.add(tileentity);
             } else {
                 this.a(tileentity);
-- 
1.8.3.msysgit.0

