From 22f920dd1e3bbe359cb06b5c5a36f9dfa38c7295 Mon Sep 17 00:00:00 2001
From: Poweruser <poweruser.rs@hotmail.com>
Date: Sat, 25 Apr 2015 19:46:36 +0200
Subject: [PATCH] Fix Null Tile Entity Worlds


diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index d92f958..8c876b5 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1878,6 +1878,7 @@ public abstract class World implements IBlockAccess {
                     }
                 }
 
+                tileentity.a(this); // MinetickMod
                 this.b.add(tileentity);
             } else {
                 this.a(tileentity);
-- 
1.8.3.msysgit.0

