From 06c1d6f57e2a6ab5a6c9be3aa6555bb9f9d28802 Mon Sep 17 00:00:00 2001
From: Poweruser <poweruser.rs@hotmail.com>
Date: Sat, 25 Apr 2015 20:01:56 +0200
Subject: [PATCH] Fix Corrupted Trapped Chest


diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 76bcb0a..0e53768 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -163,7 +163,7 @@ public class WorldServer extends World implements IAsyncTaskHandler {
         TileEntity result = super.getTileEntity(pos);
         Block type = getType(pos).getBlock();
 
-        if (type == Blocks.CHEST) {
+        if (type == Blocks.CHEST || type == Blocks.TRAPPED_CHEST) { // MinetickMod - fix trapped chests as well
             if (!(result instanceof TileEntityChest)) {
                 result = fixTileEntity(pos, type, result);
             }
-- 
1.8.3.msysgit.0

